cpp/src/strings/search/find.cu

```diff
@@ -414,6 +418,289 @@ std::unique_ptr<column> contains_warp_parallel(strings_column_view const& input,
   return results;
 }
 
+CUDF_KERNEL void multi_contains_warp_parallel_multi_scalars_fn(column_device_view const d_strings,
+                                                               column_device_view const d_targets,
+                                                               cudf::device_span<bool*> d_results)
+{
+  auto const num_targets = d_targets.size();
+  auto const num_rows    = d_strings.size();
+
+  auto const idx    = static_cast<size_type>(threadIdx.x + blockIdx.x * blockDim.x);
+  using warp_reduce = cub::WarpReduce<bool>;
+  __shared__ typename warp_reduce::TempStorage temp_storage;
+
+  if (idx >= (num_rows * cudf::detail::warp_size)) { return; }
+
+  auto const lane_idx = idx % cudf::detail::warp_size;
+  auto const str_idx  = idx / cudf::detail::warp_size;
+  if (d_strings.is_null(str_idx)) { return; }  // bitmask will set result to null.
+
+  // get the string for this warp
+  auto const d_str = d_strings.element<string_view>(str_idx);
+
+  for (size_t target_idx = 0; target_idx < num_targets; target_idx++) {
+    // Identify the target.
+    auto const d_target = d_targets.element<string_view>(target_idx);
+
+    // each thread of the warp will check just part of the string
+    auto found = false;
+    if (d_target.empty()) {
+      found = true;
+    } else {
+      for (auto i = static_cast<size_type>(lane_idx);
+           !found && ((i + d_target.size_bytes()) <= d_str.size_bytes());
+           i += cudf::detail::warp_size) {
+        // check the target matches this part of the d_str data
+        if (d_target.compare(d_str.data() + i, d_target.size_bytes()) == 0) { found = true; }
+      }
+    }
+    __syncwarp();
+    auto const result = warp_reduce(temp_storage).Reduce(found, cub::Max());
+    if (lane_idx == 0) { d_results[target_idx][str_idx] = result; }
+  }
```

Comment:
Using the outer-loop for targets seems inefficient especially for long strings since the inner-loop needs to start from the beginning of the string for each target. I think that checking for targets in the inner-loop would allow more efficient global memory access since the source string (usually longer) would only need one pass.
I apologize if you already tried that.

Reply:
If using the targets loop as a inner-loop, then require a `bool[num_targets]` warp storage instead of a `bool` and the code will be more complicated. Usually the targets are short, so global memory access should not be impacted much. Can we move this to a follow-up PR when we meet a performance issue.

Reply:
Updated, refer to this [commit](https://github.com/rapidsai/cudf/pull/16641/commits/31f4822c90dc66c92808624e592e1a8cc14c128a)